Leor/backend

la289 opened the pull request:
Backend for Dashboard Auth.
Contains:
- GET/csrf, POST/login, POST/logout endpoints
- TLS Redirect
- Expiring JWT integration with cryptographically secure signing key
- CSRF Token integration with cryptographically random tokens
- PostrgeSQL integration

Files were formatted with `go fmt`
All test cases passed with `-race` flag set

Removed password hashing utility package since it was an unnecessary abstraction layer.

la289 committed:
Password hashing and checking functionality with associated unit tests

la289 committed:
Working DBManager class. Passes all unit tests

la289 committed:
Added GetUserPassHash to DBManager. Associated unit tests pass

la289 committed:
Router, Controller, and DBManager packages. main.go and server-certs.
Controller and DBManager pass all unit tests.
Router testing has been manual until this point.

la289 committed:
used go fmt to clean up all files

la289 committed:
removed passhash util per PR feedback

la289 committed:
removed password utils per PR feedback

la289 committed:
Added unit tests for router.go. Also added some logging when requests are made

la289 committed:
added comments to tests

pierrebeaucamp commented on main.go:
```diff
@@ -0,0 +1,14 @@
+
+import (
+	"iotdashboard/router"
+	// "iotdashboard/controller"
+)
+
+func init() {
+}
```
When some things aren't needed for a particular review, it's helpful to omit them. Commented code or empty functions serve no purpose being committed. It's not a huge deal, but helps keeping things tidy and avoids errors down the line.

la289 replied:
Good Call - Resolving with next commit

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+
+func redirectTLS(w http.ResponseWriter, r *http.Request) {
+	addHeaders(w)
+	host, _, _ := net.SplitHostPort(r.Host)
```
When discarding an error, it's useful to add a comment explaining why handling the error isn't needed.

r0mant replied:
Do not ignore errors.

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+)
+
+//TODO: maybe these ports should live in a config file?
+var httpPort = ":8080"
+var httpsPort = ":9090"
```
To avoid global state, you could make those values a constant and rename them to something like `defaultHTTPport` to signal that this is only the default value.

r0mant replied:
const?

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+	"X-Frame-Options":           "DENY",
+	"X-Content-Type-Options":    "nosniff",
+	"Cache-Control":             "no-store",
+}
```
Similarly, to avoid global state, this could be a function (`GetSecHeaders()`) which returns this map (since maps can't be constants)

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+	"Cache-Control":             "no-store",
+}
+
+// Create a struct to read the email and pass from the request
+type Credentials struct {
```
For public functions or types, it's recommended to follow the GoDoc formatting for any comments. Crucially, the should start with the name of the function or type. In this case, it could read:

```
// Credentials is a struct that holds the email and password of a request
```

I won't go too much into proper documentation because that could take a lot of time, but generally, try to refrain from abbreviations and make sure that it describes the type and function in its entirety.

pierrebeaucamp replied:
On a second pass: I see you don't actually use that struct anywhere. What's its purpose?

la289 replied:
I'm using the struct as a template to decode the JSON request so that it can be broken into constituent parts

pierrebeaucamp replied:
Ah, I must have missed that 🤦. All Good

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+	mux.HandleFunc("/logout", logoutHandler)
+	mux.HandleFunc("/csrf", csrfHandler)
+
+	log.Fatal(http.ListenAndServeTLS(httpsPort, certPath, keyPath, mux))
```
Try to return errors all the way up to the `main` package before panicking. Keeping track of individual `panic` or `Fatal` statements sprinkled around the codebase can be tough over time, so aggregating them at the outermost layer is useful.

r0mant replied:
log.Fatal exits the program. Instead, return an error to the caller and let it handle it.

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+	// Get the JSON body and decode into credentials
+	err := json.NewDecoder(r.Body).Decode(&creds)
+	if err != nil {
+		http.Error(w, "Bad Request", http.StatusBadRequest)
```
Be sure to log any errors before returning.

la289 replied:
I think I added logging to all interesting locations of the Router package

pierrebeaucamp commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// TODO: turn this into middleware:
+func validateCSRF(r *http.Request, creds Credentials) (string, int) {
```
I would recommend replacing `(string, int)` with a simple `error`. You can later check what kind of error was returned:

```golang
var ErrorBadRequest = errors.New("Bad Request")

if err := validateCSRF(r, creds); err != nil {
        if err = ErrorBadRequest {
                http.Error(w, "Bad Request", http.StatusBadRequest)
                return
        }
        
        http.Error(w, "Unauthorized Request", http.StatusUnauthorized)
        return
}
```

la289 replied:
I made `validateCSRF` directly call http.Error() if CSRF validation fails instead of passing back values

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	blocklist = make(map[string]int64)
+}
```
Try to use `init` as a last resort only. For example, you could have a `Controller` struct instead, that gets initialized when needed:

```golang
type Controller struct {
        JWTkey    []byte
        PSQL      dbmanager.DBManager
        blocklist map[string]int64
}

func NewController() (Controller, error) {
 	jwtKey, err := GenerateRandomToken(256)
 	if err != nil {
 		return Controller{}, err
 	}
 	
 	PSQL, err := dbmanager.New("postgres", "myPassword", "iot_dashboard")
 	if err != nil {
 		return Controller{}, err
 	}

 	blocklist = make(map[string]int64)
 	return Controller{jwtKey, PSQL, blocklist}, nil
}
```

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+func Login(email, password string) (string, error) {
+	// use user struct instead of email and password?
+
+	// validate CSRF
```
I think you moved this into the http handler itself. You can remove these comments here

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	return bcrypt.CompareHashAndPassword(hash, []byte(password))
+}
+
+func CreateJWT(validPeriod int64) (string, error) {
```
Use `time.Duration` for `validPeriod` instead

r0mant replied:
Feels like all utility methods for creating/validating tokens can be moved to a separate package/file.

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+
+	// Create the token
+	token := jwt.NewWithClaims(jwt.GetSigningMethod("HS256"), claims)
+	// token.Claims["exp"] =
```
I assume this commented line can be removed

la289 replied:
removed both comments on line 96 and line 98 since line 97 is self explanatory

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	return tokenString, nil
+}
+
+func ValidateJWT(rawToken string) (int64, error) {
```
1. I would swap out `int64` for `time.Time` to signal that we return a timestamp, not a random number
2. It's weird to me to have a function called `ValidateJWT` return a timestamp. Calling it something like `GetJWTexpiry` might be better suited.

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+		NotBefore: time.Now().Unix(),
+		StandardClaims: jwt.StandardClaims{
+			// In JWT, the expiry time is expressed as unix milliseconds
+			ExpiresAt: time.Now().Add(time.Second * time.Duration(validPeriod)).Unix(),
```
It seems like you forgot to add `UTC()` in there

la289 replied:
@pierrebeaucamp doesn't Unix() return the number of seconds in epoch time, which is UTC?

pierrebeaucamp replied:
Yes, you're right 🙂, I should have read that more carefully. I was mostly reading back and forth and saw that you did `.UTC()` on Line 126:

```golang
now := time.Now().UTC().Unix()
```

Personally, I got into the habit of adding `.UTC()` to each `time.Now()` call so the logic is correct wether or not you later call `Unix()` on the result. But it's up to you, I would just aim to keep things consistent.

la289 replied:
I added .UTC() to these calls for readability

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	if err != nil {
+		return nil, err
+	}
+	// fmt.Println(key)
```
Leftover from debugging

la289 replied:
Oops - removing the print statement. @pierrebeaucamp do you think I should remove lines 137-139 and just do `return key, err`? This should work because is there is an error then we don't care what the key is, and if there isn't an error, then error is nil and the key is the key.

pierrebeaucamp replied:
That's personal preference. I often do just that (`return key, err`) but other team-members like to have an explicit error check in place, so it's easy to see that an error is handled.

la289 replied:
Using the return without the explicit err check

pierrebeaucamp commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	return claims.ExpiresAt, nil
+}
+
+func GenerateRandomToken(n int) ([]byte, error) {
```
I'm not 100% certain what the stated purpose of this `controller` package is, but it might not necessarily be the right place for this function. This is just me thinking out loud though, so not a huge concern.

la289 replied:
My thought process was to split the functions such that the router package handles the "web" side of things and the controller handles the "server" side of things. I think that I could remove the controller package altogether but unless there is a strong reason to, given the time constraint, I will leave it for now. Let me know if you disagree

pierrebeaucamp commented on controller/controller_test.go:
```diff
@@ -0,0 +1,79 @@
+func TestGenerateRandomString(t *testing.T) {
+	n := 5
+	s, err := GenerateRandomString(n)
+	if err != nil || len(s) < n {
```
why not `len(s) != n` ?

la289 replied:
@pierrebeaucamp because we are encoding the string to `base64`, and each element of []byte returned by `GenerateRandomToken` is 0-255, we can get elements that represent multiple characters. This can result in strings longer than n.

pierrebeaucamp replied:
The reason why I'm asking is that the input argument to `GenerateRandomString` is somewhat arbitrary. It's neither the length of the resulting string, nor is it the byte-size of the resulting string. To understand what the `5` in this scenario signals, you need to read through the implementation itself.

Not a huge deal, but I thought it's worth pointing out

la289 replied:
It is a good point. I have thrown added logic to `GenerateRandomString` that discards the part of the string longer than n characters. Test case updated accordingly

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+)
+
+//////// QUERIES
+const (
+	userExistsQuery  = `SELECT uid from users WHERE email = $1`
+	getUserInfoQuery = `SELECT * from users WHERE email = $1`
+	getUserPassQuery = `SELECT password from users WHERE email = $1`
+	createUserQuery  = `INSERT INTO users(uid,email,password) VALUES (DEFAULT, $1 , $2);`
+)
```
I'm not opposed to storing all queries as constants here, but it leads to a lot of scrolling back & forth from where the values are used. Seems ok for now though.

la289 replied:
@pierrebeaucamp would you prefer to see the raw query placed in each location where used? Easy change, definitely don't mind making it.

pierrebeaucamp replied:
It's up to you. Personally, I'd probably place it closer to where it's used, but there are probably some advantages and disadvantages either way.

I was mainly wondering if there was a strong reason to store them outside their respective function body.

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (db *DBManager) GetUserPassHash(email string) ([]byte, error) {
+	err := db.VerifyUserExists(email)
+	if err != nil {
+		return nil, err
+	}
```
Trips to the database are very expensive. You can combine `userExistsQuery` with `getUserPassQuery` in here.

la289 replied:
Leaving the query as is and using the results returned to determine if user exists

la289 replied:
Removed the `VerifyUserExists()` function from the codebase

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+	db.DB = psql
+	// only creates the table if it doesn't already exist
+	// Typically this database would exist on its own outside of docker
+	err = db.createAccountsTable()
```
in this case, it could be useful to rename the function to `createAccountsTableIfNotExists` or simply something like `initSchema`

la289 replied:
Renamed initSchemaUsers() so that we know which table it is creating

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+
+// Returns err if user is not added to DB
+func (db *DBManager) AddNewUser(email, password string) error {
+	err := db.VerifyUserExists(email)
```
Same here - this query is not needed. Because of the uniqueness constrain on the `email` column, you cannot insert a new user with the same email.

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+	userExistsQuery  = `SELECT uid from users WHERE email = $1`
+	getUserInfoQuery = `SELECT * from users WHERE email = $1`
+	getUserPassQuery = `SELECT password from users WHERE email = $1`
+	createUserQuery  = `INSERT INTO users(uid,email,password) VALUES (DEFAULT, $1 , $2);`
```
I'm not 100% sure if you need to insert `DEFAULT` into `uid`. You might be able to omit that.

la289 replied:
Tested it and you are right. Removed

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+	return nil
+}
+
+func (db *DBManager) GetUserPassHash(email string) ([]byte, error) {
```
I'm not a huge fan of calling this method `GetUserPassHash`. `GetPasswordHash` seems more descriptive.

pierrebeaucamp commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+		return err
+	}
+
+	hashedPass, err := bcrypt.GenerateFromPassword([]byte(password), 10)
+	if err != nil {
+		return err
+	}
```
I would try to do all the hashing in one package. I.e. either you can decide that it's the responsibility of the `dbmanager` to do the hashing, or the `controller`. Splitting it into two places can be overly confusing.

la289 replied:
Moved all hashing into the `dbmanager `to ensure that no unhashed passwords enter the db. Moved `CheckUserCredentials` function into `dbmanager`.

pierrebeaucamp commented on controller/controller_test.go:
```diff
@@ -0,0 +1,79 @@
+		t.Errorf("Validation succeeded when it should have failed")
+	}
+	//test expiration of token
+	time.Sleep(3 * time.Second)
```
If you need to manipulate time through tests, consider using something like https://github.com/jonboulle/clockwork to pass a clock to the component you need to test. `time.Sleep` in tests is not a good practice, although you could let it stand here if that's the only occurrence. If you decide to leave things as-is, please add a comment.

la289 replied:
I will add a TODO comment to come back to this if there is time since this is the only instance. For now leaving as is. Thanks for showing me this though, very useful

pierrebeaucamp commented on controller/controller_test.go:
```diff
@@ -0,0 +1,79 @@
+		email, pass string
+		exists      bool
+	}{
+		{"user@gmail.com", "S3cure3Pa$$", true},
```
Is this test really succeeding? Does this assume a real database is initialized with some values? If yes, consider mocking the database calls instead.

pierrebeaucamp commented on dbmanager/dbmanager_test.go:
```diff
@@ -0,0 +1,73 @@
+package dbmanager
```
Since you rely on `init` to initialize your database tests, please make sure this is a test package, so it's not included in final builds (i.e. rename this package to `dbmanager_test`)

pierrebeaucamp commented on dbmanager/dbmanager_test.go:
```diff
@@ -0,0 +1,73 @@
+		email  string
+		exists bool
+	}{
+		{"user@gmail.com", true},
```
Similar to my previous comments, I think this assumes that the database contains certain values already. Please either populate the database in the test itself, or mock the DB calls. As this stands, the test is not very useful.

pierrebeaucamp replied:
Same for the other tests in this file, I won't leave comments on each one of them

la289 replied:
All tests now use `go-sqlmock` where applicable

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+package controller
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"github.com/dgrijalva/jwt-go"
+	"iotdashboard/dbmanager"
+	"golang.org/x/crypto/bcrypt"
+	"time"
+)
```
Just as a little tip, it doesn't look like you're using `goimports` (can tell by the way these imports are organized). Check it out, it's `gofmt` on steroids and will make your life easier.

la289 replied:
Left this for last but it should have been the first thing I did. Thanks for the trick

pierrebeaucamp commented on router/router_test.go:
```diff
@@ -0,0 +1,199 @@
+		status                                  int
+	}{
+		{"POST", "/login", "user@gmail.com", "S3cure3Pa$$", "123", "123", http.StatusOK},
+		{"POST", "/login11", "user@gmail.com", "S3cure3Pa$$", "123", "123", http.StatusOK},
```
I don't understand the purpose of this test. You're testing the handler, not the router / muxer, so the different paths should make no difference.

la289 replied:
That is a really good point but the request still takes in a path, so I wanted to make sure that I wasn't doing something silly with the path in the function, for example with the `redirectTLS` function

pierrebeaucamp commented on router/router_test.go:
```diff
@@ -0,0 +1,199 @@
+
+func TestLogoutHandler(t *testing.T) {
+	token1, err := controller.CreateJWT(60)
+	token2, err := controller.CreateJWT(60)
```
This will overwrite the previous `err`, which wasn't checked yet.

pierrebeaucamp commented on router/router_test.go:
```diff
@@ -0,0 +1,199 @@
+		response := rr.Result()
+
+		//Evaluate response for status code
+		if rr.Code != c.status {
```
Ideally, you'd also test that you cannot do a second request with the same credentials, now that you're logged out.

la289 replied:
This is tested by the second case. It is using the same JWT as the first case, so we are expecting a different status response

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	"time"
+)
+
+//TODO: these probably shouldn't be global vars
```
I agree. A typical pattern in Go is you make a "service" struct/object and have all parameters required for its operation as its fields. I.e. you can have a 

```go
type DashboardController struct {
  psql   dbmanager.DBManager
  jwtKey []byte
  ...
}
```

as an example.

pierrebeaucamp commented on router/router_test.go:
```diff
@@ -0,0 +1,199 @@
+		response := rr.Result()
+
+		//Evaluate response for status code
+		if rr.Code != c.status {
```
You should also test that the redirect points towards the correct URL

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+////////Errors
+var ErrExpiredToken = errors.New("Token is expired")
+
+func init() {
```
Don't do this in init, create the service struct like explained above and init these in its constructor.

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+	blocklist = make(map[string]int64)
+}
+
+func Login(email, password string) (string, error) {
```
These can become methods on that service struct.

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,147 @@
+		return err
+	}
+	// blocklist JWT
+	blocklist[token] = exp
```
Can this map be accessed concurrently by multiple clients?

la289 replied:
Replaced map with `sync.Map` to handle concurrency

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+	createUserQuery  = `INSERT INTO users(uid,email,password) VALUES (DEFAULT, $1 , $2);`
+)
+
+//////// ERRORS
```
Please use standard Godoc format for comments.

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+
+//////// QUERIES
+const (
+	userExistsQuery  = `SELECT uid from users WHERE email = $1`
+	getUserInfoQuery = `SELECT * from users WHERE email = $1`
```
Feels like userExists query isn't really needed and getUserInfo query can be used instead.

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+		return nil, err
+	}
+
+	rows.Next()
```
Why is the return parameter ignored here?

la289 replied:
Didn't realize it had a return parameter at first. Updated with check if next row exists. If not, return `ErrUserNonexistant`

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+
+	rows.Next()
+	var hash []byte
+	rows.Scan(&hash)
```
This can return an error.

r0mant replied:
Also, `rows.Err()` isn't called.

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+
+// Returns err if user is not added to DB
+func (db *DBManager) AddNewUser(email, password string) error {
+	err := db.VerifyUserExists(email)
```
Is this check resistant to race conditions? I.e. what happens if two clients attempt to add new user with the same email at the same time.

r0mant replied:
Also, is it really needed? Why not use database to guarantee uniqueness?

la289 replied:
Removed VerifyUserExists from dbmanager and relying on query results instead

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,116 @@
+
+func (db *DBManager) createAccountsTable() error {
+	_, err := db.DB.Query(`
+		CREATE TABLE IF NOT EXISTS users(
```
Would you queries be efficient on this table?

la289 replied:
We could add indexing but that typically has a negative affect on small tables (which ours is). Is there something else I should be considering? We could add reverse lookup by domain as well to filter results but that seems unnecessary for this project.

r0mant replied:
I agree we probably wouldn't benefit much from index in this challenge scenario. Could you elaborate on the "negative effect" part please?

la289 replied:
@r0mant In general, indexing slows down write operations and is more costly to memory. It wouldn't speed up our search queries since the table is so small, so the net effect is negative.

r0mant commented on dbmanager/dbmanager_test.go:
```diff
@@ -0,0 +1,73 @@
+var PSQLerr error
+
+func init() {
+	PSQL, PSQLerr = New("postgres", "myPassword", "iot_dashboard")
```
Don't use init functions for this, initialize these in tests.

r0mant commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+	"net/http"
+)
+
+//TODO: maybe these ports should live in a config file?
```
These are fine as global constants IMO.

r0mant commented on router/router.go:
```diff
@@ -0,0 +1,182 @@
+		Value:    jwt,
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: 3,
```
What's "3"?

la289 replied:
Will use Enum `SameSiteStrict`

la289 committed:
WIP: made some of changes from PR comments

la289 committed:
WIP: incorporating PR feedback. Token util broken out of controller. Controller and Router turned into service structs

la289 committed:
WIP: PR feedback.

la289 committed:
Mocking database in all tests

la289 committed:
Exposing router service's controller parameter so that we can initialize users from main.go

la289 committed:
Main now calls router service and initializes the admin user in the DB if the user doesn't already exist

la289 committed:
Cleaned up comments and used goimports

awly commented on controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func NewController() (ControllerService, error) {
+	PSQL, err := dbmanager.New("postgres", "myPassword", "iot_dashboard")
+	if err != nil {
+		return ControllerService{}, err
+	}
+	TokenUtil, err := utils.NewTokenUtil()
```
nit: in Go, all local variables should use `lowerSnakeCase`, e.g. `psql` and `tokenUtil`

la289 replied:
@awly I'll keep an eye out for this elsewhere, but PSQL and TokenUtil are used and overwritten in outside packages, sometimes just for testing. That is why they are capitalized.

r0mant replied:
@la289 Could you clarify what you mean? These vars are local to this function.

la289 replied:
oh whoops I was looking a few lines above at the struct definition. I see now. Updated.

awly commented on controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+	if err != nil {
+		return ControllerService{}, err
+	}
+	TokenUtil, err := utils.NewTokenUtil()
```
```suggestion
	PSQL, err := dbmanager.New("postgres", "myPassword", "iot_dashboard")
	if err != nil {
		return ControllerService{}, err
	}
	TokenUtil, err := utils.NewTokenUtil()	
	if err != nil {
		return ControllerService{}, err
	}
```

awly commented on controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+
+	// create and return JWT
+	var token string
+	token, err = ct.TokenUtil.CreateJWT(60)
```
```suggestion
	token, err := ct.TokenUtil.CreateJWT(60)
```

awly commented on controller/controller_test.go:
```diff
@@ -0,0 +1,59 @@
+	"testing"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	// "fmt"
```
```suggestion
```

awly commented on controller/controller_test.go:
```diff
@@ -0,0 +1,59 @@
+func TestLoginAndLogout(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		t.Errorf("failed to open sqlmock database: %v \n", err)
```
nit: use `t.Fatalf` if you don't want the test to proceed

la289 replied:
Changed in all test setups

awly commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func (db *DBManager) getPasswordHash(email string) ([]byte, error) {
+	rows, err := db.DB.Query(`SELECT password from users WHERE email = $1`, email)
```
use `db.DB.QueryRow` when there's only 1 result

la289 replied:
Updated

awly commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,103 @@
+		return err
+	}
+
+	_, err = db.DB.Query(`INSERT INTO users(email,password) VALUES ($1 , $2);`, email, hashedPass)
```
```suggestion
	_, err = db.DB.Exec(`INSERT INTO users(email,password) VALUES ($1 , $2);`, email, hashedPass)
```

la289 replied:
added and tests updated

awly commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func (db *DBManager) initSchemaUsers() error {
+	_, err := db.DB.Query(`
```
```suggestion
	_, err := db.DB.Exec(`
```

awly commented on main.go:
```diff
@@ -0,0 +1,23 @@
+func main() {
+	router, err := router.NewRouter(":8080", ":9090")
+	if err != nil {
+		panic(err)
```
use `log.Fatal` instead of panicking.

awly commented on router/router.go:
```diff
@@ -0,0 +1,209 @@
+func (rtr *RouterService) Start(certPath, keyPath string) error {
+	log.Printf("Starting webserver ... \n")
+	//start listening for http to redirect to https
+	go http.ListenAndServe(rtr.httpPort, http.HandlerFunc(rtr.redirectTLS))
```
nit: log something if this fails

awly commented on router/router.go:
```diff
@@ -0,0 +1,209 @@
+	return RouterService{Ctrlr, httpPort, httpsPort}, nil
+}
+
+func (rtr *RouterService) Start(certPath, keyPath string) error {
```
why are key/cert arguments to `Start` and not to `NewRouter`?

la289 replied:
good point. Legacy copy/paste mistake

awly commented on router/router.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+// TODO: turn this into middleware:
+func (rtr *RouterService) validateCSRF(w http.ResponseWriter, r *http.Request, creds Credentials) {
```
return an error, so that the handler can terminate too
`http.Error` doesn't stop the handler execution

la289 replied:
added this in an earlier commit since there was a fatal error if a CSRF cookie wasn't included in the request.

awly commented on router/router.go:
```diff
@@ -0,0 +1,209 @@
+
+func (rtr *RouterService) logoutHandler(w http.ResponseWriter, r *http.Request) {
+	rtr.addHeaders(w)
+	log.Printf("Received POST/logout request")
+	if r.Method != "POST" {
+		http.Error(w, "Method not supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	var creds Credentials
+	// Get the JSON body and decode into credentials
+	err := json.NewDecoder(r.Body).Decode(&creds)
+	if err != nil {
+		log.Printf("LogoutHandler Error: %v /n", err)
+		http.Error(w, "Bad Request", http.StatusBadRequest)
+		return
+	}
```
`logoutHandler` doesn't need to validate credentials

la289 replied:
@awly This doesn't actually validate credentials. It just pulls any available parameters out of the request body. In this case just the CSRF token. Could rename the credentials struct if this is really confusing, or just add another struct with only a CSRF token in it.
For now I have added comments but if you want to see either of these changes I can make them

awly commented on router/router_test.go:
```diff
@@ -0,0 +1,238 @@
+			WillReturnRows(rows)
+
+		// Create test request
+		bodyReader := strings.NewReader(`{"email":"` + c.email +
+			`","password":"` + c.pass +
+			`","csrf":"` + c.csrfB + `"}`)
```
nit: use `fmt.Sprintf` to construct the body

la289 replied:
done

awly commented on router/router_test.go:
```diff
@@ -0,0 +1,238 @@
+		t.Errorf("Could not initialize router: %v \n", err)
+	}
+
+	token1, err1 := router.Ctrlr.TokenUtil.CreateJWT(60)
+	token2, err2 := router.Ctrlr.TokenUtil.CreateJWT(60)
+	if err1 != nil || err2 != nil {
+		t.Errorf("Failed to generate a token. Errors: \n %v \n %v \n", err1, err2)
+	}
```
handle one error at a time

awly commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+type Claims struct {
+	NotBefore int64 `json:"nbf"`
```
`jwt.StandardClaims` already has a `NotBefore` field

la289 replied:
Missed that somehow.  Good catch. Changed the structure to use the field from `jwt.StandardClaims`

awly commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,96 @@
+	claims := Claims{
+		NotBefore: time.Now().UTC().Unix(),
+		StandardClaims: jwt.StandardClaims{
+			// In JWT, the expiry time is expressed as unix milliseconds
```
```suggestion
			// In JWT, the expiry time is expressed as unix time
```

awly commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,96 @@
+	if ok {
+		return exp.(time.Time), ErrExpiredToken
+	}
+	token, err := jwt.ParseWithClaims(
+		rawToken,
+		&Claims{},
```
```suggestion
	token, err := jwt.Parse(
		rawToken,
```

la289 replied:
@awly  What is the purpose of doing this? If I don't include the claim parsing I cannot validate the nbf and expiration times, nor return an expiration time which is needed for the blacklist cleanup.

la289 committed:
Added error handling if CSRF cookie is missing

la289 committed:
Update controller/controller.go

Add error handling for TokenUtil init

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

la289 committed:
Added error handling if CSRF cookie is missing

la289 committed:
Incorporated all PR feedback up to this point

r0mant commented on controller/controller.go:
```diff
@@ -0,0 +1,54 @@
+	TokenUtil utils.TokenUtil
+}
+
+func NewController() (ControllerService, error) {
```
Why does this return a value and not a pointer?

la289 replied:
No good reason. Changed all `New____` functions to return pointers

r0mant reviewed (commented):
I see you marked many conversations resolved but I don't see the actual changes addressing them pushed?

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,93 @@
+func New(user, pass, name string) (DBManager, error) {
+	d := DBManager{dbUser: user, dbPass: pass, dbName: name}
+	err := d.connectToPSQL()
+	return d, err
```
Is there a reason to return db manager if we failed to connect?

la289 replied:
We could add an error check (previously had one)  and return nil or an empty struct but I don't think it really matters. Is there any reason not to return it as is?

r0mant commented on dbmanager/dbmanager.go:
```diff
@@ -0,0 +1,93 @@
+	if err := result.Scan(&hash); err != nil {
+		return nil, err
+	}
+	if len(hash) != 60 {
```
What's this for?

la289 replied:
You're right, can remove since the hash will be invalidated in the `CheckUserCredentials` function

r0mant commented on main.go:
```diff
@@ -0,0 +1,23 @@
+
+	err = router.Ctrlr.PSQL.AddNewUser("e@g.c", "test")
+	if err != nil {
+		log.Printf("Admin user already exists in DB")
```
Is it guaranteed that AddNewUser won't return any other error than "user already exists"? What if connection fails or something?

r0mant commented on main.go:
```diff
@@ -0,0 +1,23 @@
+
+	err = router.Start()
+	if err != nil {
+		panic(err)
```
Don't panic.

r0mant commented on router/router.go:
```diff
@@ -0,0 +1,229 @@
+		HttpOnly: true,
+		SameSite: http.SameSiteStrictMode,
+	})
+	log.Printf("Responding")
```
These logs feel like debug leftovers. They don't add much value to the logs without additional context so I would consider to either remove them or add more context so they're useful.

r0mant commented on router/router.go:
```diff
@@ -0,0 +1,229 @@
+	// CSRF Validation
+	err = rtr.validateCSRF(w, r, creds)
+	if err != nil {
+		log.Printf("CSRF Validation err: %v", err)
```
Why not log.Error? Here and everywhere else errors are logged.

la289 replied:
Not changing since the functionality doesn't exist in `log` without new packages

r0mant commented on router/router.go:
```diff
@@ -0,0 +1,229 @@
+	err = rtr.validateCSRF(w, r, creds)
+	if err != nil {
+		log.Printf("CSRF Validation err: %v", err)
+		return
```
This doesn't return anything to the client?

la289 replied:
@r0mant The response to the client is written inside of `validateCSRF`. This just logs the error and returns. I added a comment before the `log.Printf` statement to clarify

r0mant commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,95 @@
+	"github.com/dgrijalva/jwt-go"
+)
+
+type Claims struct {
```
Curious why you need this struct since it just wraps jwt.StandardClaims. I don't see any extra methods defined on Claims.

la289 replied:
You're right. Previously didn't realize that the `NotBefore` time was included in `StandardClaims`. Removed.

r0mant commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,95 @@
+			// In JWT, the expiry time is expressed as unix time
+			ExpiresAt: time.Now().UTC().Add(time.Second * validPeriod).Unix(),
+			Issuer:    "iot-dash",
+			NotBefore: time.Now().UTC().Unix(),
```
I'd make this valid sooner by a few seconds to account for skewed clocks.

la289 committed:
More PR Feedback:
- Struct Initializing functions now return pointers
- Removed extra logging messages from router
- Not panicing in main.go anymore
- Added more general error logging to AddNewUser in main.go
- Removed Claims wrapper of StandardClaims since using NotBefore from inside StandardClaims

awly commented on router/router.go:
```diff
@@ -0,0 +1,220 @@
+func (rtr *RouterService) Start() error {
+	log.Printf("Starting webserver ... \n")
+	//start listening for http to redirect to https
+	go log.Print(http.ListenAndServe(rtr.httpPort, http.HandlerFunc(rtr.redirectTLS)))
```
Without a wrapper func, this line would block until `http.ListenAndServe` returns.
When you spawn a goroutine, it evaluates all the arguments synchronously (the return error from `http.ListenAndServe` here for example) before kicking off the goroutine and proceeding.
```suggestion
	go func() {
		log.Print(http.ListenAndServe(rtr.httpPort, http.HandlerFunc(rtr.redirectTLS)))
	}()
```

awly commented on utils/tokenutils.go:
```diff
@@ -0,0 +1,89 @@
+func (tu *TokenUtil) CreateJWT(validPeriod time.Duration) (string, error) {
+	claims := jwt.StandardClaims{
+		// In JWT, the expiry time is expressed as unix time
+		ExpiresAt: time.Now().UTC().Add(time.Second * validPeriod).Unix(),
```
`validPeriod` should be a complete `time.Duration`.
Update all the callers to use the correct value.
```suggestion
		ExpiresAt: time.Now().UTC().Add(validPeriod).Unix(),
```

la289 replied:
done -> this is a good idea for user customizability

la289 committed:
Update router/router.go

Wrapping `http.ListenAndServe()` in goroutine func

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

la289 committed:
Changed CheckJWT to take in complete time.Duration, instead of converting it to seconds.

r0mant reviewed (approved):
Don't have many more comments. @awly @pierrebeaucamp what about you folks?
